File: object-oriented-programming-projects/project-8/project/HashSet.java
/**
 * HashSet is a set type container that it does not insist the same elements
 * 
 * @author Tayfun
 *
 * @param <T> parametrized container type
 */
public class HashSet <T> implements Set <T> {
	
	private T[] container;
	private int size;
	
	public HashSet(){
		container = null;
		size = 0;
	}
	
	
	public T getElem(int index){
		return container[index];
	}
	public T[] getContainer(){
		return container;
	}
	public void setContainer(T[] con){
		container = con;
	}
	public int getSize(){
		return size;
	}
	public void setSize(int sz){
		size = sz;
	}

	public Iterator<T> iterator(){
		Iterator <T> it = new Iterator<T>(this,0);
		return it;
	}


	public void add(T e){
		boolean found = false;
		for (int i=0; i<size; ++i){
			if(container[i] == e){
				found = true;
				break;
			}
		}
		if (!found){
			T[] temp = (T[]) new Object[size+1];
			for(int i = 0; i < size; ++i){
				temp[i] = container[i];
			}
			temp[size] = e;
			container = temp;
			++size;
		}
	}


	public void addAll(Collection<T> c){
		T[] tempcon = c.getContainer();
		int tempsize = c.getSize();
		for (int i = 0; i < tempsize; ++i){
			boolean found = false;
			for (int j = 0; j < size; ++j){
				if(tempcon[i] == container[j]){
					found = true;
					break;
				}
			}
			if(!found){
				T[] newcon =(T[]) new Object[size+1];
				for(int k=0; k<size; ++k)
					newcon[k] = container[k];
				newcon[size] = tempcon[i];
				container = newcon;
				++size;
			}
		}
	}


	public void clear(){
		size = 0;
	}


	public boolean contains(T e){
		boolean found = false;
		for (int i=0; i<size; ++i){
			if(container[i] == e){
				found = true;
				break;
			}
		}
		return found;
	}


	public boolean containsAll(Collection<T> c){
		T[] tempcon = c.getContainer();
		int tempsize = c.getSize();
		for (int i=0; i<tempsize; ++i){
			boolean found = false;
			for (int j=0; j<size; ++j){
				if(tempcon[i] == container[j]){
					found = true;
					break;
				}
			}
			if(!found)
				return false;
		}
		return true;
	}


	public boolean isEmpty(){
		return (size == 0);
	}


	public void remove(T e){
		for (int i=0; i<size; ++i){
			if(container[i] == e){
				T[] temp = (T[]) new Object[size-1];
				for(int j=0; j<i; ++j){
					temp[j] = container[j];
				}
				for(int j=i; j<size-1; ++j){
					temp[j] = container[j+1];
				}
				container = temp;
				--size;
				return;
			}
		}
		return;
	}


	public void removeAll(Collection<T> c){
		T[] tempcon = c.getContainer();
		int tempsize = c.getSize();
		for (int i=0; i<tempsize; ++i){
			for (int j=0; j<size; ++j){
				if (tempcon[i] == container[j]){
					T[] temp = (T[]) new Object[size-1];
					for(int k=0; k<j; ++k){
						temp[k] = container[k];
					}
					for(int k=j; k<size-1; ++k){
						temp[k] = container[k+1];
					}
					container = temp;
					--size;
					break;
				}
			}
		}
	}


	public void retainAll(Collection<T> c){
		T[] tempcon = c.getContainer();
		int tempsize = c.getSize();
		for (int i=0; i<size; ++i){
			boolean found = false;
			for (int j=0; j<tempsize; ++j){
				if (container[i] == tempcon[j]){
					found = true;
					break;
				}
			}
			if(!found){
				T[] temp = (T[]) new Object[size-1];
				for(int j=0; j<i; ++j){
					temp[j] = container[j];
				}
				for(int j=i; j<size-1; ++j){
					temp[j] = container[j+1];
				}
				container = temp;
				--size;
				--i;
			}
		}
	}


	public int size(){
		return size;
	}
	
	
}

File: object-oriented-programming-projects/project-6/project/hashset.cpp


namespace SpaceCollection
{
template<class E, class C>
Iterator<E,C> HashSet<E,C>::iterator(){
	Iterator<E,C> it;
	it.iterator = container.begin();
	it.size = container.size();
	it.index = 0;
	it.conPtr = &container;
	return it;
}

template<class E, class C>
void HashSet<E,C>::add(E e){
	bool found = false;
	for (auto i = container.begin(); i != container.end(); ++i){
		if(*i == e){
			found = true;
			break;
		}
	}
	if (!found){
		auto j = container.end();
		container.insert(j,e);
	}
}

template<class E, class C>
void HashSet<E,C>::addAll(Collection<E,C> & c){
	HashSet<E,C> * hsp = dynamic_cast<HashSet<E,C> *>(&c);
	for (auto i = hsp->container.begin(); i != hsp->container.end(); ++i){
		bool found = false;
		for (auto j = container.begin(); j != container.end(); ++j){
			if(*i == *j){
				found = true;
				break;
			}
		}
		if(!found){
			auto k = container.end();
			container.insert(k,*i);
		}
	}
}

template<class E, class C>
void HashSet<E,C>::clear(){
	container.clear();
}

template<class E, class C>
bool HashSet<E,C>::contains(E e){
	bool found = false;
	for (auto i = container.begin(); i != container.end(); ++i){
		if(*i == e){
			found = true;
			break;
		}
	}
	return found;
}

template<class E, class C>
bool HashSet<E,C>::containsAll(Collection<E,C> & c){
	HashSet<E,C> * hsp = dynamic_cast<HashSet<E,C> *>(&c);
	for (auto i = hsp->container.begin(); i != hsp->container.end(); ++i){
		int found = false;
		for (auto j = container.begin(); j != container.end(); ++j){
			if(*i == *j){
				found = true;
				break;
			}
		}
		if(!found)
			return false;
	}
	return true;
}

template<class E, class C>
bool HashSet<E,C>::isEmpty(){
	return (container.size() == 0);
}

template<class E, class C>
void HashSet<E,C>::remove(E e){
	for (auto i = container.begin(); i != container.end(); ++i){
		if(*i == e){
			container.erase(i);
			return;
		}
	}
	return;
}

template<class E, class C>
void HashSet<E,C>::removeAll(Collection<E,C> & c){
	HashSet<E,C> * hsp = dynamic_cast<HashSet<E,C> *>(&c);
	for (auto i = hsp->container.begin(); i != hsp->container.end(); ++i){
		for (auto j = container.begin(); j != container.end(); ++j){
			if (*i == *j){
				container.erase(j);
				break;
			}
		}
	}
}

template<class E, class C>
void HashSet<E,C>::retainAll(Collection<E,C> & c){
	HashSet<E,C> * hsp = dynamic_cast<HashSet<E,C> *>(&c);
	for (auto i = container.begin(); i != container.end(); ++i){
		bool found = false;
		for (auto j = hsp->container.begin(); j != hsp->container.end(); ++j){
			if (*i == *j){
				found = true;
				break;
			}
		}
		if(!found){
			container.erase(i);
			--i;
		}
	}
}

template<class E, class C>
int HashSet<E,C>::size(){
	return container.size();
}
//	end of SpaceCollection namespace
}